File: collection/CPICRec_abacavir_HLA-B/recommendation.js
const drug = 'abacavir';

const reference = {
  'HLA-B': { field: 'diplotype', value: '57:01' }
};

const keysuffix = {
  'HLA-B': { negative: 'noncarrier', positive: 'carrier' }
};

const recommendations = {
  'hla-b57:01noncarrier': {
    implication: 'Low or reduced risk of abacavir hypersensitivity',
    recommendation: 'abacavir: Use abacavir per standard dosing guidelines',
    classification: 'Strong'
  },
  'hla-b57:01carrier': {
    implication: 'Significantly increased risk of abacavir hypersensitivity',
    recommendation: 'abacavir: Abacavir is not recommended',
    classification: 'Strong'
  }
};

function dosingrecommendation(inputs) {
  try {
    const genes = {};
    const output = {};
    let searchkeyReady = true;
    let searchKey = '';
    const lowercaseInput = {};

    for (const inputkey in inputs) {
      const key = inputkey.toLowerCase();
      lowercaseInput[key] = inputs[inputkey];
    }

    for (const genekey in reference) {
      const key = genekey.toLowerCase();
      if (!lowercaseInput[key]) break;

      genes[genekey] = {
        diplotype: lowercaseInput[key].diplotype || '',
        phenotype: (lowercaseInput[key].phenotype || '').toLowerCase()
      };

      const targetfield = reference[genekey].field;
      searchkeyReady = searchkeyReady && genes[genekey][targetfield] !== '';

      if (targetfield === 'diplotype') {
        if (genes[genekey].diplotype.includes(reference[genekey].value)) {
          searchKey += key + reference[genekey].value + keysuffix[genekey].positive;
        } else {
          searchKey += key + reference[genekey].value + keysuffix[genekey].negative;
        }
      }

      if (targetfield === 'phenotype') {
        if (genes[genekey].phenotype !== '') {
          searchKey += key + genes[genekey].phenotype.replace('metabolizer', '').replace(/\s/g, '');
        }
      }
    }

    if (searchkeyReady) {
      const recommendation = recommendations[searchKey];
      if (recommendation != null) {
        output.type = 'CPIC Recommendation';
        output.drug = drug;
        output.genes = structuredClone(genes);
        output.recommendation = {
          implication: recommendation.implication,
          content: recommendation.recommendation,
          classification: recommendation.classification
        };
        return output;
      } else {
        return `Incorrect/invalid input for drug ${drug}`;
      }
    } else {
      return 'Incorrect/invalid input.';
    }
  } catch (error) {
    return 'Error: ' + error;
  }
}

export { dosingrecommendation };

File: collection/CPICRec_codeine_CYP2D6/cpicrec_codeine_cyp2d6/recommendation.py
import copy

drug = "Codeine"
reference = {"CYP2D6": {"field": "phenotype", "value": ""}}
keysuffix = {"CYP2D6": {"positive": "", "negative": ""}}  # adjust if needed
recommendations = {
    "cyp2d6ultrarapid": {
        "implication": "Increased formation of morphine following codeine administration, leading to higher risk of toxicity",
        "recommendation": "Avoid codeine use due to potential for toxicity. Alternatives that are not affected by this CYP2D6 phenotype include morphine and nonopioid analgesics. Tramadol and, to a lesser extent, hydrocodone and oxycodone are not good alternatives because their metabolism is affected by CYP2D6 activity",
        "classification": "Strong",
    },
    "cyp2d6normal": {
        "implication": "Normal morphine formation",
        "recommendation": "Use label-recommended age- or weight-specific dosing.",
        "classification": "Strong",
    },
    "cyp2d6intermediate": {
        "implication": "Reduced morphine formation",
        "recommendation": "Use label-recommended age- or weight-specific dosing. If no response, consider alternative analgesics such as morphine or a nonopioid. Monitor tramadol use for response.",
        "classification": "Moderate",
    },
    "cyp2d6poor": {
        "implication": "Greatly reduced morphine formation following codeine administration, leading to insufficient pain relief.",
        "recommendation": "Avoid codeine use due to lack of efficacy. Alternatives that are not affected by this CYP2D6 phenotype include morphine and nonopioid analgesics. Tramadol and, to a lesser extent, hydrocodone and oxycodone are not good alternatives because their metabolism is affected by CYP2D6 activity; these agents should be avoided",
        "classification": "Strong",
    },
}


def dosingrecommendation(inputs):
    try:
        genes = {}
        output = {}
        searchkey_ready = True
        lowercase_input = {}
        search_key = ""

        # convert input keys to lowercase
        for inputkey, value in inputs.items():
            lowercase_input[inputkey.lower()] = value

        for genekey, ref in reference.items():
            key = genekey.lower()
            if key not in lowercase_input:
                break

            genes[genekey] = {}
            genes[genekey]["diplotype"] = lowercase_input[key].get("diplotype", "")
            genes[genekey]["phenotype"] = lowercase_input[key].get("phenotype", "").lower()

            targetfield = ref["field"]
            searchkey_ready = searchkey_ready and genes[genekey].get(targetfield, "") != ""

            if targetfield == "diplotype":
                if ref["value"] in genes[genekey]["diplotype"]:
                    search_key += key + ref["value"] + keysuffix[genekey]["positive"]
                else:
                    search_key += key + ref["value"] + keysuffix[genekey]["negative"]

            if targetfield == "phenotype":
                if genes[genekey]["phenotype"] != "":
                    pheno = genes[genekey]["phenotype"].replace("metabolizer", "").replace(" ", "")
                    search_key += key + pheno

        if searchkey_ready:
            if search_key in recommendations:
                output["type"] = "CPIC Recommendation"
                output["drug"] = drug
                output["genes"] = copy.deepcopy(genes)
                rec = recommendations[search_key]
                output["recommendation"] = {
                    "implication": rec["implication"],
                    "content": rec["recommendation"],
                    "classification": rec["classification"],
                }
                return output
            else:
                return f"Incorrect/invalid input for drug {drug}"
        else:
            return "Incorrect/invalid input."

    except Exception as error:
        return f"Error: {error}"

